### PR TITLE
Start kotlinpoet-km artifact

### PR DESCRIPTION
**Pull request by ZacSweers**
Part of #707. No tests right now yet pending feedback on the implementation and also thoughts on what tests for this should look like!

### CODE REVIEW

File: kotlinpoet-km/gradle.properties [hunk 1]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=kotlinpoet-km
```
Comment (JakeWharton) [inline kotlinpoet-km/gradle.properties]: Don't like this name here or in the package or the type names. Took a while to figure out what it meant.

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+ * instance of its values.
+ */
+@KotlinPoetKm
+fun KotlinClassMetadata.Class.toImmutableKmClass(): ImmutableKmClass =
+    toKmClass().immutable()
+
+/** @return an immutable representation of this [KmClass]. */
+@KotlinPoetKm
+@JvmName("immutableOf")
+fun KmClass.immutable(): ImmutableKmClass {
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `toImmutable()`?

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+
+/**
+ * Immutable representation of [KmClass].
+ *
+ * Represents a Kotlin class.
+ *
+ * @property flags Class flags, consisting of [Flag.HAS_ANNOTATIONS], visibility flag, modality flag and [Flag.Class] flags.
+ * @property name Name of the class.
+ * @property typeParameters Type parameters of the class.
+ * @property supertypes Supertypes of the class. The first element is the superclass (or [Any])
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+  )
+}
+
+/**
+ * Immutable representation of [KmPackage].
+ *
+ * Represents a Kotlin package fragment, including single file facades and multi-file class parts.
+ *
+ * @property functions Functions in the package fragment.
+ * @property properties properties in the package fragment.
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `properties Properties`

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+ * @property typeAliases typeAliases in the package fragment.
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `typeAliases Type aliases`

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 6]
```diff
@@ -0,0 +1,887 @@
...
+      signature = signature
+  )
+}
+
+/**
+ * Immutable representation of [KmConstructor].
+ *
+ * Represents a constructor of a Kotlin class.
+ *
+ * @property flags constructor flags, consisting of [Flag.HAS_ANNOTATIONS], visibility flag and [Flag.Constructor] flags
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Constructor`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 7]
```diff
@@ -0,0 +1,887 @@
...
+      lambdaClassOriginName
+  )
+}
+
+/**
+ * Immutable representation of [KmFunction].
+ *
+ * Represents a Kotlin function declaration.
+ *
+ * @property flags function flags, consisting of [Flag.HAS_ANNOTATIONS], visibility flag, modality flag and [Flag.Function] flags
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Function`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 8]
```diff
@@ -0,0 +1,887 @@
...
+ * @property name the name of the function
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The name`

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 9]
```diff
@@ -0,0 +1,887 @@
...
+      syntheticMethodForAnnotations
+  )
+}
+
+/**
+ * Immutable representation of [KmProperty].
+ *
+ * Represents a Kotlin property declaration.
+ *
+ * @property flags property flags, consisting of [Flag.HAS_ANNOTATIONS], visibility flag, modality flag and [Flag.Property] flags
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Property flags`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 10]
```diff
@@ -0,0 +1,887 @@
...
+ * @property name the name of the property
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The name`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 11]
```diff
@@ -0,0 +1,887 @@
...
+ * @property getterFlags property accessor flags, consisting of [Flag.HAS_ANNOTATIONS], visibility flag, modality flag
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Property accessor`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 12]
```diff
@@ -0,0 +1,887 @@
...
+ *   and [Flag.PropertyAccessor] flags
+ * @property setterFlags property accessor flags, consisting of [Flag.HAS_ANNOTATIONS], visibility flag, modality flag
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Property accessor`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 13]
```diff
@@ -0,0 +1,887 @@
...
+      versionRequirements.map { it.immutable() }
+  )
+}
+
+/**
+ * Immutable representation of [KmTypeAlias].
+ *
+ * Represents a Kotlin type alias declaration.
+ *
+ * @property flags type alias flags, consisting of [Flag.HAS_ANNOTATIONS] and visibility flag
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Type alias`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 14]
```diff
@@ -0,0 +1,887 @@
...
+ * @property name the name of the type alias
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The name`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 15]
```diff
@@ -0,0 +1,887 @@
...
+      varargElementType?.immutable()
+  )
+}
+
+/**
+ * Immutable representation of [KmValueParameter].
+ *
+ * Represents a value parameter of a Kotlin constructor, function or property setter.
+ *
+ * @property flags value parameter flags, consisting of [Flag.ValueParameter] flags
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Value parameter`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 16]
```diff
@@ -0,0 +1,887 @@
...
+ * @property name the name of the value parameter
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The name`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 17]
```diff
@@ -0,0 +1,887 @@
...
+      annotations.unmodifiable()
+  )
+}
+
+/**
+ * Immutable representation of [KmTypeParameter].
+ *
+ * Represents a type parameter of a Kotlin class, function, property or type alias.
+ *
+ * @property flags type parameter flags, consisting of [Flag.TypeParameter] flags
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Type parameter`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 18]
```diff
@@ -0,0 +1,887 @@
...
+ * @property name the name of the type parameter
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The name`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 19]
```diff
@@ -0,0 +1,887 @@
...
+ * @property id the id of the type parameter, useful to be able to uniquely identify the type parameter in different contexts where
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The id`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 20]
```diff
@@ -0,0 +1,887 @@
...
+ *           the name isn't enough (e.g. `class A<T> { fun <T> foo(t: T) }`)
+ * @property variance the declaration-site variance of the type parameter
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The declaration`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 21]
```diff
@@ -0,0 +1,887 @@
...
+      annotations.unmodifiable()
+  )
+}
+
+/**
+ * Immutable representation of [KmType].
+ *
+ * Represents a type.
+ *
+ * @property flags type flags, consisting of [Flag.Type] flags
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Type flags`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 22]
```diff
@@ -0,0 +1,887 @@
...
+ */
+@KotlinPoetKm
+data class ImmutableKmType internal constructor(
+    override val flags: Flags,
+    val classifier: KmClassifier,
+    val arguments: List<ImmutableKmTypeProjection>,
+    /**
+     * Abbreviation of this type. Note that all types are expanded for metadata produced by the Kotlin compiler. For example:
+     *
+     *     typealias A<T> = MutableList<T>
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: enclose snippet in ```?

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 23]
```diff
@@ -0,0 +1,887 @@
...
+     *
+     *     fun foo(a: A<Any>) {}
+     *
+     * The type of the `foo`'s parameter in the metadata is actually `MutableList<Any>`, and its abbreviation is `A<Any>`.
+     */
+    val abbreviatedType: ImmutableKmType?,
+    /**
+     * Outer type of this type, if this type's classifier is an inner class. For example:
+     *
+     *     class A<T> { inner class B<U> }
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: enclose snippet in ```?

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 24]
```diff
@@ -0,0 +1,887 @@
...
+     * Flexible types in Kotlin include platform types in Kotlin/JVM and `dynamic` type in Kotlin/JS.
+     */
+    val flexibleTypeUpperBound: ImmutableKmFlexibleTypeUpperBound?,
+    val isRaw: Boolean,
+    val annotations: List<KmAnnotation>
+) : ImmutableKmWithFlags {
+  /**
+   * `true` if this is an extension type (i.e. String.() -> Unit vs (String) -> Unit).
+   *
+   * See details: https://discuss.kotlinlang.org/t/announcing-kotlinx-metadata-jvm-library-for-reading-modifying-metadata-of-kotlin-jvm-class-files/7980/27?u=hzsweers
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: remove `u=hzsweers`?

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 25]
```diff
@@ -0,0 +1,887 @@
...
+
+/**
+ * Immutable representation of [KmEffect].
+ *
+ * Represents an effect (a part of the contract of a Kotlin function).
+ *
+ * Contracts are an internal feature of the standard Kotlin library, and their behavior and/or binary format
+ * may change in a subsequent release.
+ *
+ * @property type type of the effect
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Type of`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 26]
```diff
@@ -0,0 +1,887 @@
...
+ * @property invocationKind optional number of invocations of the lambda parameter of this function,
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Optional`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 27]
```diff
@@ -0,0 +1,887 @@
...
+  return ImmutableKmTypeProjection(variance, type?.immutable())
+}
+
+/**
+ * Immutable representation of [KmTypeProjection].
+ *
+ * Represents type projection used in a type argument of the type based on a class or on a type alias.
+ * For example, in `MutableMap<in String?, *>`, `in String?` is the type projection which is the first type argument of the type.
+ *
+ * @property variance the variance of the type projection, or `null` if this is a star projection
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The variance`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 28]
```diff
@@ -0,0 +1,887 @@
...
+ * @property type the projected type, or `null` if this is a star projection
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `The projected`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 29]
```diff
@@ -0,0 +1,887 @@
...
+fun KmFlexibleTypeUpperBound.immutable(): ImmutableKmFlexibleTypeUpperBound {
+  return ImmutableKmFlexibleTypeUpperBound(type.immutable(), typeFlexibilityId)
+}
+
+/**
+ * Immutable representation of [KmFlexibleTypeUpperBound].
+ *
+ * Represents an upper bound of a flexible Kotlin type.
+ *
+ * @property type upper bound of the flexible type
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Upper`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 30]
```diff
@@ -0,0 +1,887 @@
...
+ * @property typeFlexibilityId id of the kind of flexibility this type has. For example, "kotlin.jvm.PlatformType" for JVM platform types,
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: `Id of`, `.` at the end of the line

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 31]
```diff
@@ -0,0 +1,887 @@
...
+ *                          or "kotlin.DynamicType" for JS dynamic type
+ */
+@KotlinPoetKm
+data class ImmutableKmFlexibleTypeUpperBound internal constructor(val type: ImmutableKmType,
```
Comment (Egorand) [inline kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: nit: reformat constructor to one param per line

File: kotlinpoet-km/gradle.properties (hunk 1 above)
Comment (ZacSweers) [reply kotlinpoet-km/gradle.properties]: Honestly I don't know what a good option is. I'd be fine with `kotlinpoet-metadata`, but also would be curious to know if we can figure out the name separately as a followup to reduce churn from the implementation

Comment (ZacSweers) [reply kotlinpoet-km/gradle.properties]: I do feel strongly about them being in the type names though I think, as the real types are prefixed with that and these are just immutable versions of them

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt (hunk 22 above)
Comment (ZacSweers) [reply kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: It's markdown, so this is a code block with the indention. Was copied from the kotlinx-metadata doc

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/KotlinPoetKm.kt [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+@KotlinPoetKm
+fun Class<*>.asImmutableKmClass(): ImmutableKmClass = readMetadata(::getAnnotation).asImmutableKmClass()
+/** @return a new [ImmutableKmClass] representation of the Kotlin metadata for [this] type. */
+@KotlinPoetKm
+fun TypeElement.asImmutableKmClass(): ImmutableKmClass = readMetadata(::getAnnotation).asImmutableKmClass()
+
+@KotlinPoetKm
+fun Metadata.asImmutableKmClass(): ImmutableKmClass {
+  return when (val metadata = readKotlinClassMetadata()) {
+    //  return when (metadata) {
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/KotlinPoetKm.kt]: 3cf732b

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt (hunk 24 above)
Comment (ZacSweers) [reply kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: 3cf732b

File: kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt [hunk 33]
```diff
@@ -0,0 +1,887 @@
...
+    val typeFlexibilityId: String?) {
+  fun mutable(): KmFlexibleTypeUpperBound {
+    return KmFlexibleTypeUpperBound(type.mutable(), typeFlexibilityId)
+  }
+}
+
+private fun <E> List<E>.unmodifiable(): List<E> = Collections.unmodifiableList(this)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply kotlinpoet-km/src/main/kotlin/com/squareup/kotlinpoet/km/immutableNodes.kt]: Went with `toImmutableList()` to match Util's name 2f5233e

**Comment by ZacSweers** [comment]
Commenting here rather than inline
- mass migrated all `mutable()`/`immutable()` names to `asMutable()`/`asImmutable()`. Using `as` to match convention in other APIs here
- mass fixed grammar nits
- ran spotless to fix all the formatting

**Comment by ZacSweers** [comment]
Reading the reasoning on https://github.com/square/okhttp/pull/5187, I understand the as/to distinction better. Will switch them around here

**Comment by Egorand** [comment]
Merging this to unblock further development. Question regarding the artifact name is still open.

**Pull request merged**